pkg/distributor/distributor.go

```diff
@@ -688,7 +685,57 @@ func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
 			d.nonHASamples.WithLabelValues(userID).Add(float64(numSamples))
 		}
 
-		cleanupInDefer = false
+		dontCleanup()
+		return next(ctx, req, cleanup)
+	}
+}
+
+func (d *Distributor) prePushRelabelMiddleware(next push.Func) push.Func {
+	return func(ctx context.Context, req *mimirpb.WriteRequest, cleanup func()) (*mimirpb.WriteResponse, error) {
+		conditionalCleanup, dontCleanup := getConditionalCleanup(cleanup)
+		defer conditionalCleanup()
+
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		var removeTsIndexes []int
+		for tsIdx := 0; tsIdx < len(req.Timeseries); tsIdx++ {
+			ts := req.Timeseries[tsIdx]
+
+			if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
+				l := relabel.Process(mimirpb.FromLabelAdaptersToLabels(ts.Labels), mrc...)
+				ts.Labels = mimirpb.FromLabelsToLabelAdapters(l)
+			}
+
+			for _, labelName := range d.limits.DropLabels(userID) {
+				removeLabel(labelName, &ts.Labels)
+			}
+
+			if len(ts.Labels) == 0 {
+				if len(removeTsIndexes) == 0 {
+					// If we have to add one index into this slice then there is a good chance that we'll have to add more.
```

Comment:
How do we know that? Series are sent in batches, using shards in Prometheus. Series in the same shard may not be related at all, so we could be allocating unnecessarily big slice here.

Reply:
I assumed that many series produced by one exporter usually have similar or even equal labels and they get scraped together in a single scrape, if drop_label/relabel rules ended up dropping every single label in a series then it's likely that they would do the same with other series of the same exporter because they likely have the same or a similar labels.

Maybe allocating a slice with then capacity of `len(req.Timeseries)` is a bit much, but any smaller number would just be a guess about how much might possibly be enough. What do you think we should do? Not pre-allocate the slice at all, or pre-allocate some small factor of `len(req.Timeseries)`, like `len(req.Timeseries)/4` or something like that?

Reply:
We don't use label relabeling in distributors ourselves, so it's not going to affect us too much either way.

Labels here include `__name__` label too, so if number of labels is zero, it means that the whole series is being dropped. I would say that's rare occurence, so starting with empty or small (eg. 10 elements) slice would work just fine here.

Reply:
ok, let's start empty then. 
any number like `10` is a pure guess in my opinion, in that case we might as well rely on Go to be smart enough to grow the slice in reasonable steps.

Reply:
updated ae4b9ad542e0f06e2f6958d148dd981e80f67366